# Clean data and analyze this used car listing. 

#### This dataset was originally created by a user in Kaggle, but it is not available anymore. The original dataset can be found in: https://data.world/data-society/used-cars-data, but the dataset was modified so this analysis would make sense. The modified dataset can be found in this GitHub repository named "autos". 

#### The data dictionary provided with data is as follows:

dateCrawled - When this ad was first crawled. All field-values are taken from this date.

name - Name of the car.

seller - Whether the seller is private or a dealer.

offerType - The type of listing

price - The price on the ad to sell the car.

abtest - Whether the listing is included in an A/B test.

vehicleType - The vehicle Type.

yearOfRegistration - The year in which the car was first registered.

gearbox - The transmission type.

powerPS - The power of the car in PS.

model - The car model name.

kilometer - How many kilometers the car has driven.

monthOfRegistration - The month in which the car was first registered.

fuelType - What type of fuel the car uses.

brand - The brand of the car.

notRepairedDamage - If the car has a damage which is not yet repaired.

dateCreated - The date on which the eBay listing was created.

nrOfPictures - The number of pictures in the ad.

postalCode - The postal code for the location of the vehicle.

lastSeenOnline - When the crawler saw this ad last online.

In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings("ignore")

In [2]:
autos = pd.read_csv ("C://Users//johng//Desktop//data//autos.csv", encoding="Windows-1252")

In [3]:
# info and 5 columns
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

#### Converting columns names from camelcase to snakecase and reword some to make more sense.

In [4]:
mapping_dict = {
"dateCrawled": "date_crawled",
"offerType": "offer_type",
"vehicleType": "vehicle_type",
"yearOfRegistration": "registration_year",
"powerPS": "power_ps",
"monthOfRegistration": "registration_month",
"fuelType": "fuel_type",
"notRepairedDamage": "unrepaired_damage",
"dateCreated": "ad_created",
"nrOfPictures": "nr_of_pictures",
"postalCode": "postal_code",
"lastSeen": "last_seen"
}

autos = autos.rename(columns=mapping_dict )
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [5]:
autos.describe(include='all')
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

#### Seller and offertype have the same values. These columns will be droped.
#### Price column and odometer are not numeric. "Odemeter" is going to be called "odometer_km". 

In [6]:
# just run this cell once.
autos.drop(["seller","offer_type"],axis=1,inplace=True)

In [7]:
#price
autos["price"] = (autos["price"]
                  .str.replace("$","")
                  .str.replace(",","")
                  .astype(float)
                )


#odometer
autos["odometer"] = (autos["odometer"].
                     str.replace("km","").
                     str.replace(",","").
                     astype(int)
                    )

autos.rename({'odometer': 'odometer_km'},axis=1,inplace=True)

In [8]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0  5000.0  control          bus               2004    manuell       158   
1  8500.0  control    limousine               1997  automatik       286   
2  8990.0     test    limousine               2009    manuell       102   
3  4350.0  control   kleinwagen               2007  automatik        71   
4  1350.0     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
0              nein  2016-03-26 00:00:00               0        79588   
1              nein  2016-04-04 00:00:00               0        71034   
2              nein  2016-03-26 00:00:00               0        35394   
3              nein  2016-03-12 00:00:00               0        33729   
4              nein  2016-04-01 00:00:00               0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

In [9]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

#### Used cars range between 650 and 200,000 dollars.

In [10]:
autos["price"] = autos.loc[autos["price"].between(650, 200000), "price"]

In [11]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

#### Study of the correlation between price and odometer:

In [12]:
column_1 = autos["price"]
column_2 = autos["odometer_km"]
correlation = column_1. corr(column_2)
print(correlation)

-0.4186593422709814


#### This correlation shows that the price increases the odometer decreases, which is correct.

In [13]:
# INVESTIGATING THE DATE_CRAWLED COLUMN
# .mul(100) is to put the value in percent under 100
date_crawled = autos["date_crawled"].str[:10].value_counts(normalize= True, dropna = False).mul(100).round(1).astype(str) + '%'
date_crawled.sort_index()

2016-03-05    2.5%
2016-03-06    1.4%
2016-03-07    3.6%
2016-03-08    3.3%
2016-03-09    3.3%
2016-03-10    3.2%
2016-03-11    3.2%
2016-03-12    3.7%
2016-03-13    1.6%
2016-03-14    3.7%
2016-03-15    3.4%
2016-03-16    2.9%
2016-03-17    3.2%
2016-03-18    1.3%
2016-03-19    3.5%
2016-03-20    3.8%
2016-03-21    3.8%
2016-03-22    3.3%
2016-03-23    3.2%
2016-03-24    2.9%
2016-03-25    3.2%
2016-03-26    3.2%
2016-03-27    3.1%
2016-03-28    3.5%
2016-03-29    3.4%
2016-03-30    3.4%
2016-03-31    3.2%
2016-04-01    3.4%
2016-04-02    3.5%
2016-04-03    3.9%
2016-04-04    3.7%
2016-04-05    1.3%
2016-04-06    0.3%
2016-04-07    0.1%
Name: date_crawled, dtype: object

In [14]:
autos["registration_year"].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

#### That registration year = 9999 does not make sense. 25% of data is from 1999. Every add created in "date_crawled" is from 2016, that is the reason for the maximum date:

In [15]:
autos["registration_year"] = autos.loc[autos["registration_year"].between(1999, 2016), "registration_year"]

In [16]:
autos["registration_year"].value_counts(normalize=True).mul(100)

2000.0    9.028021
2005.0    8.115529
1999.0    8.075153
2004.0    7.367231
2003.0    7.340314
2006.0    7.289171
2001.0    7.275713
2002.0    6.818121
2007.0    6.201717
2008.0    6.005222
2009.0    5.647223
2011.0    4.398267
2010.0    4.298673
2012.0    3.561142
2016.0    3.542300
2013.0    2.169524
2014.0    1.792684
2015.0    1.073995
Name: registration_year, dtype: float64

#### The majority adds were registered in 2000.

#### Analyze the brands vs price vs mileage

In [17]:
autos['brand'].value_counts(dropna=True)

volkswagen        10687
opel               5461
bmw                5429
mercedes_benz      4734
audi               4283
ford               3479
renault            2404
peugeot            1456
fiat               1308
seat                941
skoda               786
mazda               757
nissan              754
smart               701
citroen             701
toyota              617
sonstige_autos      546
hyundai             488
volvo               457
mini                424
mitsubishi          406
honda               399
kia                 356
alfa_romeo          329
porsche             294
suzuki              293
chevrolet           283
chrysler            181
dacia               129
daihatsu            128
jeep                110
subaru              109
land_rover           99
saab                 80
daewoo               79
trabant              78
jaguar               77
rover                69
lancia               57
lada                 31
Name: brand, dtype: int64

In [18]:
#selecting the top 20 brands  
brands_20 = autos['brand'].value_counts().head(20).index
brands_20

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'smart',
       'citroen', 'toyota', 'sonstige_autos', 'hyundai', 'volvo', 'mini'],
      dtype='object')

In [19]:
brand_price ={}

for brand in brands_20:    
    mean_price = round(autos.loc[autos["brand"] == brand, "price"].mean())
    brand_price[brand] = mean_price

In [20]:
brand_price

{'volkswagen': 5955,
 'opel': 3548,
 'bmw': 8596,
 'mercedes_benz': 8811,
 'audi': 9702,
 'ford': 4572,
 'renault': 2965,
 'peugeot': 3522,
 'fiat': 3421,
 'seat': 4874,
 'skoda': 6636,
 'mazda': 4665,
 'nissan': 5446,
 'smart': 3583,
 'citroen': 4145,
 'toyota': 5308,
 'sonstige_autos': 12860,
 'hyundai': 5859,
 'volvo': 5477,
 'mini': 10617}

In [21]:
brand_mileage = {}

for brand in brands_20:    
    mean_mileage = round(autos.loc[autos["brand"] == brand, "odometer_km"].mean())
    brand_mileage[brand] = mean_mileage

In [22]:
brand_mileage

{'volkswagen': 128955,
 'opel': 129299,
 'bmw': 132522,
 'mercedes_benz': 130886,
 'audi': 129644,
 'ford': 124132,
 'renault': 128224,
 'peugeot': 127352,
 'fiat': 117037,
 'seat': 122062,
 'skoda': 110948,
 'mazda': 125132,
 'nissan': 118979,
 'smart': 100756,
 'citroen': 119765,
 'toyota': 115989,
 'sonstige_autos': 87189,
 'hyundai': 106783,
 'volvo': 138632,
 'mini': 89375}

#### Create pandas series from price and mileage dictionaries:

In [23]:
brand_price_series = pd.Series(brand_price)
brand_price_series

volkswagen         5955
opel               3548
bmw                8596
mercedes_benz      8811
audi               9702
ford               4572
renault            2965
peugeot            3522
fiat               3421
seat               4874
skoda              6636
mazda              4665
nissan             5446
smart              3583
citroen            4145
toyota             5308
sonstige_autos    12860
hyundai            5859
volvo              5477
mini              10617
dtype: int64

In [24]:
brand_mileage_series = pd.Series(brand_mileage)
brand_mileage_series

volkswagen        128955
opel              129299
bmw               132522
mercedes_benz     130886
audi              129644
ford              124132
renault           128224
peugeot           127352
fiat              117037
seat              122062
skoda             110948
mazda             125132
nissan            118979
smart             100756
citroen           119765
toyota            115989
sonstige_autos     87189
hyundai           106783
volvo             138632
mini               89375
dtype: int64

#### Create a dataframe from price and mileage series:

In [25]:
price_df = pd.DataFrame(brand_price_series, columns=['mean_price'])

In [26]:
price_df["mean_mileage"] = brand_mileage_series 

#### Analyze both mean price and mean mileage in one dataframe:

In [32]:
df = price_df.sort_values("mean_mileage")

In [33]:
print(df)

                mean_price  mean_mileage
sonstige_autos       12860         87189
mini                 10617         89375
smart                 3583        100756
hyundai               5859        106783
skoda                 6636        110948
toyota                5308        115989
fiat                  3421        117037
nissan                5446        118979
citroen               4145        119765
seat                  4874        122062
ford                  4572        124132
mazda                 4665        125132
peugeot               3522        127352
renault               2965        128224
volkswagen            5955        128955
opel                  3548        129299
audi                  9702        129644
mercedes_benz         8811        130886
bmw                   8596        132522
volvo                 5477        138632


In [36]:
df["diff"] = df["mean_mileage"] - df["mean_price"]

In [37]:
print(df)

                mean_price  mean_mileage    diff
sonstige_autos       12860         87189   74329
mini                 10617         89375   78758
smart                 3583        100756   97173
hyundai               5859        106783  100924
skoda                 6636        110948  104312
toyota                5308        115989  110681
fiat                  3421        117037  113616
nissan                5446        118979  113533
citroen               4145        119765  115620
seat                  4874        122062  117188
ford                  4572        124132  119560
mazda                 4665        125132  120467
peugeot               3522        127352  123830
renault               2965        128224  125259
volkswagen            5955        128955  123000
opel                  3548        129299  125751
audi                  9702        129644  119942
mercedes_benz         8811        130886  122075
bmw                   8596        132522  123926
volvo               

#### It is logical that with higher mean price there is a lower mileage. Although, if we compare smart with 3.583 mean price
#### and 100.756 mileage, should have a higher price than the next model that is hyundai with 5.859 mean price and 106.783 mileage.
#### There are more columns in this dataset that can explain this difference like vehicle type, gearbox and unrepaired damage.